plugins/link/SharedMemory.cpp

```diff
@@ -15,12 +17,43 @@
 #	include <errno.h>
 #endif
 
+#include <atomic>
+
+#include <climits>
+#include <cstdint>
 #include <cstring>
 
 #include <iostream>
 
+// A Chunk is one atomically accessed chunk of shared memory.
+// The only way to check at compile time if operations on a
+// given atomic type are always lock-free before C++17 is
+// using the C macros ATOMIC_*_LOCK_FREE, which aren't defined
+// for fixed-width integer types.
+// int32_t is virtually always the same as int, long, or both,
+// so we check for those cases.
+#if INT_MAX == INT32_MAX
+using Chunk = int;
+static_assert(ATOMIC_INT_LOCK_FREE == 2, "32-bit atomic signed integers aren't always lock-free");
+#elif LONG_MAX == INT32_MAX
+using Chunk = long;
+static_assert(ATOMIC_LONG_LOCK_FREE == 2, "32-bit atomic signed integers aren't always lock-free");
+#else
+#	error "INT32_MAX isn't the same as INT_MAX or LONG_MAX"
+#endif
```

Comment:
`is_always_lock_free` would be ideal to avoid making it a runtime error, and it requires C++17. I thought I'd read somewhere that Mumble is C++14, but maybe I imagined it. I'll switch to `is_always_lock_free`, since the macro stuff sure isn't great!

Reply:
Ah yeah - I forgot that Mumble is stuck with C++17 :roll_eyes: 
But I think it should be fine to static assert on `std::atomic< std::uint32_t >::is_lock_free`. If it does require some locks for unaligned access, then so be it - we don't expect that to happen on typical hardware that Mumble runs on anyway.
If you are feeling adventurous, you could add a macro switch checking the C++ standard and use the `always_lock_free` function instead, if C++17 is active (then it'd automatically switch once we switch to using C++17)

Reply:
I've applied the second idea. It still checks at compile time but is a bit less unsightly and no longer adds a preprocessor conditional that might never be triggered.

Reply:
yeah, your second suggestion sounds good. We can worry about the other cases as soon as we receive a report of someone complaining about it ^^